### PR TITLE
Eng/dqm

### PR DESCRIPTION
**Pull request by basiav**

### CODE REVIEW

**Comment by basiav** [comment]
There are 2 input data sources in the demos:
- karate club  - from networkx lib, so it is not stored in any file
-  brain graph data - stored in [demo/network_community_detection/brain_networks_data](https://github.com/qc-lab/QHyper/tree/eng/dqm/demo/network_community_detection/brain_networks_data).

If you find the demo dir too messy, we can:
- delete the [demo/network_community_detection](https://github.com/qc-lab/QHyper/tree/eng/dqm/demo/network_community_detection) dir, then the [community_detection_brain.ipynb](https://github.com/qc-lab/QHyper/blob/eng/dqm/demo/community_detection_brain.ipynb) would probably be to delete as well as it relies on the data and functions from this dir

and

- keep only [community_detection_karate_club.ipynb](https://github.com/qc-lab/QHyper/blob/eng/dqm/demo/community_detection_karate_club.ipynb) as a community detection demo

File: QHyper/solvers/gurobi/gurobi.py [hunk 1]
```diff
@@ -22,11 +20,11 @@ class Gurobi(Solver):  # todo works only for quadratic expressions
     def __init__(self, problem: Problem) -> None:
         self.problem: Problem = problem
 
-    def solve(
-            self,
-            params_inits: dict[str, Any]
-    ) -> Any:
-        gpm = gp.Model("name")
+    def solve(self, params_inits: dict[str, Any]) -> Any:
+        name = params_inits["name"] if "name" in params_inits else "name"
```
Comment (tmek1244) [inline QHyper/solvers/gurobi/gurobi.py:24]:
```suggestion
        name = params_inits.get("name", "name")
```

File: QHyper/solvers/dqm/__init__.py [hunk 2]
```diff
@@ -0,0 +1 @@
+from .dqm import DQM
```
Comment (tmek1244) [inline QHyper/solvers/dqm/__init__.py:1]:
```suggestion
from .dqm import DQM

```

Comment (tmek1244) [reply QHyper/solvers/dqm/__init__.py:1]: It's a convention to have a new line at the at of file

File: QHyper/solvers/dqm/dqm.py [hunk 3]
```diff
@@ -0,0 +1,23 @@
+import os
+from typing import Any
+
+from dwave.system import LeapHybridDQMSampler
+from QHyper.problems.base import Problem
+from QHyper.solvers.base import Solver
+from QHyper.solvers.converter import Converter
+
+token = os.environ["DWAVE_API_TOKEN"]
```
Comment (tmek1244) [inline QHyper/solvers/dqm/dqm.py:9]: Global variables should be in uppercase
```suggestion
DWAVE_API_TOKEN = os.environ["DWAVE_API_TOKEN"]
```

File: QHyper/solvers/dqm/dqm.py [hunk 4]
```diff
@@ -0,0 +1,23 @@
...
+
+class DQM(Solver):
+    def __init__(self, problem: Problem, time: float) -> None:
+        self.problem: Problem = problem
+        self.time: float = time
+
+    def solve(self, params_inits: dict[str, Any] = {}) -> Any:
+        converter = Converter()
+        dqm = converter.to_dqm(self.problem)
+        sampler = LeapHybridDQMSampler(token=token)
```
Comment (tmek1244) [inline QHyper/solvers/dqm/dqm.py:20]:
```suggestion
        sampler = LeapHybridDQMSampler(token=DWAVE_API_TOKEN)
```

File: QHyper/solvers/cqm/cqm.py [hunk 5]
```diff
@@ -8,6 +10,9 @@
 from QHyper.optimizers.base import Optimizer
 
 
+token = os.environ['DWAVE_API_TOKEN']
```
Comment (tmek1244) [inline QHyper/solvers/cqm/cqm.py:13]:
```suggestion
DWAVE_API_TOKEN = os.environ['DWAVE_API_TOKEN']
```

File: QHyper/solvers/cqm/cqm.py [hunk 6]
```diff
@@ -51,7 +56,7 @@ def solve(self, params_inits: dict[str, Any] = {}) -> Any:
 
         converter = Converter()
         cqm = converter.to_cqm(self.problem)
-        sampler = LeapHybridCQMSampler()
+        sampler = LeapHybridCQMSampler(token=token)
```
Comment (tmek1244) [inline QHyper/solvers/cqm/cqm.py:59]:
```suggestion
        sampler = LeapHybridCQMSampler(token=DWAVE_API_TOKEN)
```

File: QHyper/problems/maxcut.py [hunk 7]
```diff
@@ -1,27 +1,31 @@
...
-            [f'x{i}' for i in range(max(v for edge in edges for v in edge) + 1)]
-        ))
+        self.variables = sympy.symbols(
+            " ".join(
+                [
+                    f"x{i}"
+                    for i in range(max(v for edge in edges for v in edge))
+                ]
+            )
+        )
```
Comment (tmek1244) [inline QHyper/problems/maxcut.py:19]: To be fair, I'm not a fan of this version which occupies 8 lines

File: QHyper/hyperparameter_gen/parser.py [hunk 8]
```diff
@@ -123,18 +123,32 @@ def tmp(left: Any, right: Any, multiplier: int = 1) -> list[Any]:
...
+            self.polynomial: sympy.core.Expr | None = equation
+            self.dictionary: dict = self.as_dict()
+        elif isinstance(equation, dict):
+            self.polynomial: sympy.core.Expr | None = None
+            self.dictionary: dict = equation
+        else:
+            raise Exception(
+                f"Expression equation must be an instance of \
+                    sympy.core.Expr or dict, got of type: \
+                        {type(equation)} instead"
```
Comment (tmek1244) [inline QHyper/hyperparameter_gen/parser.py:137]: Don't use \ to split lines . There are many better ways to do that 
```suggestion
                "Expression equation must be an instance of "
                "sympy.core.Expr or dict, got of type: "
                f"{type(equation)} instead"
```

File: QHyper/problems/community_detection.py [hunk 9]
```diff
@@ -0,0 +1,193 @@
...
+    objective_function : Expression
+        objective function in SymPy syntax
+    constraints : list[Expression]
+        list of problem constraints in SymPy syntax, (default [])
+    variables : int
+        number of qubits in the circuit, equal to number of nodes
+        in the graph
+    """
+
+    def __init__(self, network_data: Network, N_communities: int = 2) -> None:
```
Comment (tmek1244) [inline QHyper/problems/community_detection.py:80]: I don't like starting variable name with capital letter, and imo this N doesn't add anything
```suggestion
    def __init__(self, network_data: Network, communities: int = 2) -> None:
```

File: QHyper/problems/community_detection.py [hunk 10]
```diff
@@ -0,0 +1,193 @@
...
+        self.variables: tuple[
+            sympy.Symbol
+        ] = self._encode_discretes_to_one_hots()
+        self._set_objective_function()
+        self._set_one_hot_constraints(N_communities)
+
+    def _encode_discrete_to_one_hot(
+        self, discrete_variable: sympy.Symbol, case_value: int
+    ) -> sympy.Symbol:
+        discrete_id = int(str(discrete_variable)[len("x") :])
```
Comment (tmek1244) [inline QHyper/problems/community_detection.py:107]: Isn't len("x") always one? XD

File: QHyper/problems/community_detection.py [hunk 11]
```diff
@@ -0,0 +1,193 @@
...
+    def _encode_discretes_to_one_hots(self) -> tuple[sympy.Symbol]:
+        one_hots: tuple[sympy.Symbol] = sympy.symbols(
+            " ".join(
+                [
+                    str(self._encode_discrete_to_one_hot(var, case_val))
+                    for var in self._get_discrete_variable_representation()
+                    for case_val in range(self.cases)
+                ]
+            )
+        )
```
Comment (tmek1244) [inline QHyper/problems/community_detection.py:120]:
```suggestion
        one_hots: tuple[sympy.Symbol] = sympy.symbols( " ".join([
                str(self._encode_discrete_to_one_hot(var, case_val))
                for var in self._get_discrete_variable_representation()
                for case_val in range(self.cases)
        ]))
```

File: QHyper/problems/community_detection.py [hunk 12]
```diff
@@ -0,0 +1,193 @@
...
+                    x_i, x_j = sympy.symbols(f"x{i}"), sympy.symbols(f"x{j}")
+                    s_i = str(self._encode_discrete_to_one_hot(x_i, case_val))
+                    s_j = str(self._encode_discrete_to_one_hot(x_j, case_val))
+                    equation[(s_i, s_j)] = self.B[i, j]
+
+        equation = {key: -1 * val for key, val in equation.items()}
+
+        self.objective_function = Expression(equation)
+
+    def _set_one_hot_constraints(self, N_communities: int) -> None:
```
Comment (tmek1244) [inline QHyper/problems/community_detection.py:148]: same as before 
```suggestion
    def _set_one_hot_constraints(self, communities: int) -> None:
```

File: QHyper/problems/community_detection.py [hunk 13]
```diff
@@ -0,0 +1,193 @@
...
+                expression += dummy
+            expression += ONE_HOT_CONST
+            self.constraints.append(Expression(expression))
+
+    def decode_solution(self, solution: dict) -> dict:
+        ONE_HOT_VALUE = 1.0
+        decoded_solution: dict = {}
+
+        for variable, value in solution.items():
+            _, id = variable[0], int(variable[len("s") :])
```
Comment (tmek1244) [inline QHyper/problems/community_detection.py:170]: That also should work right? xd And again len("s") is always 1
```suggestion
            id = int(variable[len("s") :])
```

File: QHyper/problems/community_detection.py [hunk 14]
```diff
@@ -0,0 +1,193 @@
...
+            if value == ONE_HOT_VALUE:
+                case_value = id % self.cases
+                variable_id = id // self.cases
+                decoded_solution[variable_id] = case_value
+
+        decoded_solution = self.sort_decoded_solution(decoded_solution)
+        return decoded_solution
```
Comment (tmek1244) [inline QHyper/problems/community_detection.py:177]:
```suggestion
        return self.sort_decoded_solution(decoded_solution)
```

File: QHyper/problems/community_detection.py [hunk 15]
```diff
@@ -0,0 +1,193 @@
...
+
+    def sort_encoded_solution(self, encoded_solution: dict) -> dict:
+        keyorder = self.variables
+        solution_by_keyorder: dict = {
+            str(k): encoded_solution[str(k)]
+            for k in keyorder
+            if str(k) in encoded_solution
+        }
+        return solution_by_keyorder
```
Comment (tmek1244) [inline QHyper/problems/community_detection.py:186]:
```suggestion
        return {
            str(k): encoded_solution[str(k)]
            for k in self.variables
            if str(k) in encoded_solution
        }
```

File: QHyper/problems/community_detection.py [hunk 16]
```diff
@@ -0,0 +1,193 @@
...
+
+    def sort_decoded_solution(self, decoded_solution: dict) -> dict:
+        keyorder = [int(str(v)[len("x") :]) for v in self.variables]
```
Comment (tmek1244) [inline QHyper/problems/community_detection.py:189]: same here with len("x")

File: QHyper/problems/community_detection.py [hunk 17]
```diff
@@ -0,0 +1,193 @@
...
+        decoded_solution_by_keyorder: dict = {
+            k: decoded_solution[k] for k in keyorder if k in decoded_solution
+        }
+        return decoded_solution_by_keyorder
```
Comment (tmek1244) [inline QHyper/problems/community_detection.py:193]:
```suggestion
        return {
            k: decoded_solution[k] for k in keyorder if k in decoded_solution
        }
```

File: QHyper/solvers/converter.py [hunk 18]
```diff
@@ -64,3 +71,45 @@ def to_qubo(problem: Problem) -> tuple[QUBO, float]:
...
         return cast(tuple[QUBO, float], bqm.to_qubo())  # (qubo, offset)
+
+    @staticmethod
+    def to_dqm(problem: Problem) -> DiscreteQuadraticModel:
+        dqm = dimod.DiscreteQuadraticModel()
+
+        BIN_OFFSET = 1 if problem.cases == 1 else 0
+
+        def binary_to_discrete(v: sympy.Symbol) -> sympy.Symbol:
+            id = int(str(v)[len("s") :])
```
Comment (tmek1244) [inline QHyper/solvers/converter.py:82]:
```suggestion
            id = int(str(v)[1:])
```

File: QHyper/problems/maxcut.py (hunk 7 above)
Comment (basiav) [reply QHyper/problems/maxcut.py]: Oops black got too rigid

**Pull request merged**